skynet/dgpu/network.py

```diff
@@ -344,3 +286,158 @@ async def get_input_data(self, ipfs_hash: str) -> Image:
+
+    @property
+    def poll_index(self) -> int:
+        return self._poll_index
+
+    async def _fetch_results(self):
+        self._results = await self._conn.find_results()
+
+    async def _fetch_statuses_for_id(self, rid: int):
+        self._status_by_rid[rid] = await self._conn.get_status_by_request_id(rid)
+
+    async def update_state(self):
+        '''
+        Get a "snapshot" of current contract table state
+
+        '''
+        # raw queue from chain
+        _queue = await self._conn.get_work_requests_last_hour()
+
+        # filter out invalids
+        self._queue = []
+        for req in _queue:
+            try:
+                msgspec.json.decode(req.body, type=BodyV0)
+                self._queue.append(req)
+
+            except msgspec.ValidationError:
+                logging.exception(f'dropping req {req.id} due to:')
+                ...
+
+        random.shuffle(self._queue)
+        self._queue = sorted(
+            self._queue,
+            key=lambda req: convert_reward_to_int(req.reward),
+            reverse=True
+        )
+
+        async with trio.open_nursery() as n:
+            n.start_soon(self._fetch_results)
+            for req in self._queue:
+                n.start_soon(
+                    self._fetch_statuses_for_id, req.id)
+
+
+        maybe_update_tui(lambda tui: tui.network_update(self))
+
+    async def wait_data_update(self):
+        await self._new_data.wait()
+
+    async def _state_update_task(self, poll_time: float):
+        '''
+        Long running task, polls gpu contract tables latest table rows,
+        awakes any self._data_event waiters
+
+        '''
+        while True:
+            start_time = time.time()
+            await self.update_state()
+            elapsed = time.time() - start_time
+            self._new_data.set()
+            await trio.sleep(max(poll_time - elapsed, 0.1))
+            self._new_data = trio.Event()
+            self._poll_index += 1
+
+    # views into data
+
+    @property
+    def queue_len(self) -> int:
+        return len(self._queue)
+
+    @property
+    def first(self) -> RequestV0 | None:
+        if len(self._queue) > 0:
+            return self._queue[0]
+
+        else:
+            return None
+
+    def competitors_for_id(self, request_id: int) -> set[str]:
+        return set((
+            status.worker
+            for status in self._status_by_rid[request_id]
+            if status.worker != self._conn.config.account
+        ))
+
+    # predicates
+
+    def is_request_filled(self, request_id: int) -> bool:
+        return request_id in [
+            result.request_id for result in self._results
+        ]
+
+    def is_request_in_progress(self, request_id: int) -> bool:
+        return request_id in self._status_by_rid
+
+    def should_compete_for_id(self, request_id: int) -> bool:
+        return bool(
+            self._conn.config.non_compete &
+            self.competitors_for_id(request_id)
+        )
```

Comment:
with a name like `should_compete_for_id` shouldn't this return true if we want to compete?

Reply:
we would need to update `daemon.py` if this changes

Reply:
yup good catch adding a `not` in front of the bool cast